contracts/oracle/src/contract/feed.rs

```diff
@@ -127,20 +149,442 @@ pub fn try_query_alarms<OracleBase>(
 where
     OracleBase: Currency + DeserializeOwned,
 {
     let total_registered = Feeders::total_registered(storage)?;
     use crate::state::config::Config as OracleConfig;
     let config = OracleConfig::load(storage)?;
     let oracle = Feeds::<OracleBase>::with(config.price_config);
-    oracle.calc_all_prices(storage, block_time, total_registered)
+    oracle.all_prices_iter(storage, tree.swap_pairs_df(), block_time, total_registered)
+}
+
+struct LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency,
+    Querier: PriceQuerier,
+{
+    price_querier: Querier,
+    stack: Vec<BasePrice<SwapGroup, OracleBase>>,
+    err: bool,
+}
+
+impl<Querier, OracleBase> LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency,
+    Querier: PriceQuerier,
+{
+    // TODO: improve implementation
+    fn recover<B, Q>(&mut self) -> bool
+    where
+        B: Currency,
+        Q: Currency,
+    {
+        if Q::TICKER == OracleBase::TICKER {
+            self.stack.clear();
+            self.err = false;
+        } else {
+            let parent_idx = self
+                .stack
+                .iter()
+                .position(|parent| parent.base_ticker() == Q::TICKER);
+
+            if let Some(n) = parent_idx {
+                self.stack.truncate(n + 1);
+                self.err = false;
+            } else {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl<OracleBase, Querier> AnyVisitorPair for &mut LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency + DeserializeOwned,
+    Querier: PriceQuerier,
+{
+    type Output = Option<BasePrice<SwapGroup, OracleBase>>;
+    type Error = ContractError;
+
+    fn on<B, Q>(self) -> Result<Self::Output, Self::Error>
+    where
+        B: Currency + DeserializeOwned,
+        Q: Currency + DeserializeOwned,
+    {
+        // recovery mode
+        if self.err && !self.recover::<B, Q>() {
+            return Ok(None);
+        }
+
+        let price: Option<BasePrice<SwapGroup, OracleBase>> = loop {
+            match self
+                .stack
+                .last()
+                .map(TryInto::<Price<Q, OracleBase>>::try_into)
+            {
+                None => {
+                    debug_assert_eq!(Q::TICKER, OracleBase::TICKER);
+
+                    break self.price_querier.price::<B, OracleBase>()?;
+                }
+                Some(Ok(price_parent)) => {
+                    break self
+                        .price_querier
+                        .price::<B, Q>()?
+                        .map(|price| price * price_parent)
+                }
+                _ => {
+                    self.stack.truncate(self.stack.len() - 1);
+                }
+            }
+        }
+        .map(|price| {
+            let bprice: BasePrice<SwapGroup, OracleBase> = price.into();
+            self.stack.push(bprice.clone());
+            bprice
+        });
```

Comment:
Too cryptic logic and hard to maintain further, therefore:
- please replace `loop-break` with `Iterator::find_map` that returns Option<(position, price)>
- None means no price
- Some means the next price is that and it is an ancestor of the stack element at that position
- eliminate `LegCmd::err`
- eliminate `LegCmd::recover`
- remove any logic related to setting/reading `err`